lib/rules/prefer-arrow-callback.js

```diff
@@ -246,7 +249,37 @@ module.exports = {
                     !scopeInfo.super &&
                     !scopeInfo.meta
                 ) {
-                    context.report(node, "Unexpected function expression.");
+                    context.report({
+                        node,
+                        message: "Unexpected function expression.",
+                        fix(fixer) {
+
+                            // FIXME: Is there a better way to detect parser support for arrow functions?
+                            if (!context.parserOptions || !context.parserOptions.ecmaVersion || context.parserOptions.ecmaVersion < 6) {
+
+                                // If ES6 parsing is not enabled, creating arrow functions will cause syntax errors, so don't do that.
+                                return null;
+                            }
+
+                            if (!callbackInfo.isLexicalThis && scopeInfo.this) {
+
+                                // If the callback function does not have .bind(this) and contains a reference to `this`, there
+                                // is no way to determine what `this` should be, so don't perform any fixes.
+                                return null;
+                            }
+
+                            const params = "(" + node.params.map(param => sourceCode.getText(param)).join(", ") + ")";
```

Comment:
Oh, good catch!

We can use `context.getDeclaredVariables(node)` to check duplicated names.
If the node is a function node, the method returns the function name (in this case, nothing) and parameters.
It would work for the following code as well:

    foo(function({a, b = 1, c: [d, a]}) { })

Example:

```js
function isDuplicatedParameter(variable) {
    return variable.defs.filter(d => d.type === "Parameter").length >= 2;
}
const parameters = context.getDeclaredVariables(node);
const hasDuplicated = parameters.some(isDuplicatedParameter);
```